File: LeetCodePractice/src/Math/MinimumIncrementToMakeArrayUnique.java
package Math;

import java.util.Arrays;

/*Given an array of integers A, a move consists of choosing any A[i], and incrementing it by 1.
Return the least number of moves to make every value in A unique.

Example 1:

Input: [1,2,2]
Output: 1
Explanation:  After 1 move, the array could be [1, 2, 3].
Example 2:

Input: [3,2,1,2,1,7]
Output: 6
Explanation:  After 6 moves, the array could be [3, 4, 1, 2, 5, 7].
It can be shown with 5 or less moves that it is impossible for the array to have all unique values.
 
Note:
0 <= A.length <= 40000
0 <= A[i] < 40000*/
public class MinimumIncrementToMakeArrayUnique {
	public int minIncrementForUnique(int[] A) {
        int ans = 0;
        if(A == null || A.length == 0){
            return ans;
        }
        int[] map = new int[10000];
        for(int a : A){
        	map[a]++;
        }
        int duplicate = 0;
        for(int x = 0; x < 10000; x++){
        	if(map[x] > 1){
        		duplicate += map[x] - 1;
        		ans -= x *(map[x] - 1);
        	} else if(map[x] == 0 && duplicate > 0){
        		duplicate--;
        		ans += x;
        	}
        }
        return ans;
    }
	
	public int minIncrementForUniqueBySort(int[] A) {
		int ans = 0;
        if(A == null || A.length == 0){
            return ans;
        }
        Arrays.sort(A);
        int pre = A[0];
        for(int i  = 1; i < A.length; i++){
        	if(A[i] < pre + 1){
        		ans += pre + 1 - A[i];
                A[i] = pre + 1;
        	} 
        	pre = A[i];
        }
        return ans;
    }
}

File: LeetCodePython/MinimumIncrementToMakeArrayUnique.py
"""
You are given an integer array nums. In one move, you can pick an index i where 0 <= i < nums.length and increment nums[i] by 1.

Return the minimum number of moves to make every value in nums unique.

The test cases are generated so that the answer fits in a 32-bit integer.

Example 1:
Input: nums = [1,2,2]
Output: 1
Explanation: After 1 move, the array could be [1, 2, 3].

Example 2:
Input: nums = [3,2,1,2,1,7]
Output: 6
Explanation: After 6 moves, the array could be [3, 4, 1, 2, 5, 7].
It can be shown with 5 or less moves that it is impossible for the array to have all unique values.

Constraints:
1 <= nums.length <= 10^5
0 <= nums[i] <= 10^5

analysis:
Sort
TC: O(NlogN)
"""
from typing import List


class MinimumIncrementToMakeArrayUnique:
    def minIncrementForUnique(self, nums: List[int]) -> int:
        res = 0
        nums.sort()
        for i in range(1, len(nums)):
            if nums[i] <= nums[i - 1]:
                res += nums[i - 1] + 1 - nums[i]
                nums[i] = nums[i - 1] + 1
        return res

    def minIncrementForUnique2(self, nums: List[int]) -> int:
        res = 0
        most = max(nums)
        freq = [0] * (most + len(nums) + 1)
        for num in nums:
            freq[num] += 1
        for i, f in enumerate(freq):
            if f >= 1:
                dup = f - 1
                freq[i + 1] += dup
                freq[i] = 1
                res += dup
        return res

